File: 0414/main.cpp
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        priority_queue<int, vector<int>, greater<int>> maxs;
        unordered_set<int> sets;
        for(int& n : nums) {
            if(maxs.size() < 3 && sets.count(n)==0) {
                maxs.push(n);
                sets.insert(n);
            } else {
                if(n > maxs.top() && sets.count(n)==0) {
                    sets.erase(maxs.top());
                    sets.insert(n);
                    maxs.pop();
                    maxs.push(n);
                }
            }
        }
        if(sets.size()==1) return nums[0];
        if(sets.size()==2) {
            maxs.pop();
            return maxs.top();
        }
        return maxs.top();
    }
};

File: 0414/main.rs
use std::collections::{HashSet, BinaryHeap};
use std::cmp::{max, Reverse};

impl Solution {
    pub fn third_max(nums: Vec<i32>) -> i32 {
        let mut sets: HashSet<i32> = HashSet::new();
        let mut maxs: BinaryHeap<Reverse<i32>> = BinaryHeap::new();

        for n in nums.into_iter() {
            if(!sets.contains(&n)) {
                if maxs.len() < 3  {
                    sets.insert(n);
                    maxs.push(Reverse(n));
                } else if n > maxs.peek().unwrap().0 {
                    sets.remove(&n);
                    maxs.pop();
                    sets.insert(n);
                    maxs.push(Reverse(n));
                }
            }
        }
        if sets.len()==2 {
            maxs.pop();
        }
        maxs.peek().unwrap().0
    }
}